### PR TITLE
Branch names  always escape

### PR DESCRIPTION
**Pull request by danielbeardsley**
* Deal with slashes in branch names
* Git module: use shellescape on all commands

### CODE REVIEW

**Comment by davidrans** [comment]
CR ☕️

**Comment by djmetzle** [comment]
CR :diamond_shape_with_a_dot_inside:

**Comment by k0rvusk0r4x** [comment]
Not a clue how to test this. Ideas?

**Comment by danielbeardsley** [comment]
Addison found a way. Something like this:

```
cd ~/
git clone ...
cd ...
bundle install
cd ~/Code
~/git-scripts/bin/feature switch branch/with/slashes
```

**Comment by k0rvusk0r4x** [comment]
Addison and I began working on this. Thus far we have a bit of an update to the commands but are still working out testing.

Update to the commands:

1. `cd ~/`
2. `git clone ...`
3. `cd ...`
4. `bundle install --path vendor/bundle`
5. `bundle install`
6. Check out branch
7. `cd ~/Code`
8. `alias feature="~/git-scripts/bin/feature"`
9. `~/git-scripts/bin/feature switch branch/with/slashes`

**Comment by addison-grant** [comment]
I think I've narrowed it down to 
```ruby
when 'pull'                                                                
   Git::run_safe(["git fetch"])                                            
                                                                           
   current = Git::current_branch                                           
   upstream = "#{current}@{upstream}"                                      
   upstream_hash = Git::branch_hash(upstream)                              
                                                                           
   if upstream_hash == ''                                                  
      die "Your branch #{current} hasn't been pushed, nothing to pull from"
   end                                                                     
```
I think this statement is causing an empty string:
```
upstream = "#{current}@{upstream}" 
```

**Comment by addison-grant** [comment]
Replacing it with
```ruby
upstream = `git rev-parse #{current}@{upstream}`.strip
```
seems to work.

**Comment by k0rvusk0r4x** [comment]
1. `cd ~/`
2. `git clone ...`
3. `cd ...`
4. `bundle install --path vendor/bundle`
5. `bundle install`
6. Check out branch
7. `cd ~/Code`
8. `alias feature="~/git-scripts/bin/feature"`
9. `~/git-scripts/bin/feature switch branch/with/slashes`
10. `feature pull`
11. `feature merge`
12. After the merge process, `git reset --hard origin/master` because we don't want to deploy yet.

This all now works with no issues.

QA 😺

File: bin/feature [hunk 1]
```diff
@@ -339,8 +339,7 @@ when 'pull'
    Git::run_safe(["git fetch"])
 
    current = Git::current_branch
-   upstream = "#{current}@{upstream}"
-   upstream_hash = Git::branch_hash(upstream)
+   upstream_hash = `git rev-parse #{current}@{upstream}`.strip
```
Comment (addison-grant) [inline bin/feature]: I think this doesn't do everything that `Git::branch_hash` does..

dev_block :sunny: on maybe reverting this particular change.

**Comment by addison-grant** [comment]
dev_block ☀️ on maybe reverting this particular change: https://github.com/iFixit/git-scripts/pull/154#discussion_r298662054.

File: lib/git.rb [hunk 2]
```diff
@@ -123,12 +123,12 @@ def self.delete_current_branch()
...
       Git::switch_branch(devBranch)
 
-      `git branch -d #{branch}`.strip
+      `git branch -d #{branch.shellescape}`.strip
    end
 
    # Returns the SHA1 hash that the specified branch or symbol points to
    def self.branch_hash(branch)
-      `git rev-parse --verify --quiet "#{branch}" 2>/dev/null`.strip
+      `git rev-parse --verify --quiet "#{branch.shellescape}" 2>/dev/null`.strip
```
Comment (addison-grant) [inline lib/git.rb]: This is escaping the `{` and `}` in `"#{current}@{upstream}"`

Comment (addison-grant) [reply lib/git.rb]: which causes feature pull to fail. Do we want another function called upstream_branch_hash that only escapes the branch name?

**Comment by addison-grant** [comment]
un_dev_block :sunny:

**Comment by djmetzle** [comment]
CR :gear:

**Comment by k0rvusk0r4x** [comment]
1. `cd ~/`
2. `git clone ...`
3. `cd ...`
4. `bundle install --path vendor/bundle`
5. Check out branch
6. `cd ~/Code`
7. `alias feature="~/git-scripts/bin/feature"`
8. `feature switch branch/with/slashes`
9. `feature pull`
10. `feature merge`
11. After the merge process, `git reset --hard origin/master` because we don't want to deploy yet.
12. `unalias feature`

Still working with no issues.

QA 😺

**Comment by addison-grant** [comment]
@danielbeardsley, should this file be updated?

https://github.com/iFixit/git-scripts/blob/branch-names--always-escape/git-scripts.gemspec#L6

It says 0.8.0, but the rubygems page lists the latest version as 0.8.1

https://rubygems.org/gems/git-scripts/versions/0.8.1

**Comment by addison-grant** [comment]
Should we bump it up to 0.8.2? I am guessing we're going to be publishing this and then reinstalling it.

**Comment by danielbeardsley** [comment]
> Should we bump it up to 0.8.2?

Yes please!  Maybe even 0.9!

**Comment by danielbeardsley** [comment]
CR 👍

**Comment by addison-grant** [comment]
Also, out of all my repos, `feature pull` only breaks in git-scripts
```
>> feature pull                                                                    
Traceback (most recent call last):
        14: from /usr/local/bin/feature:23:in `<main>'
        13: from /usr/local/bin/feature:23:in `load'
        12: from /usr/local/share/gems/gems/git-scripts-0.8.1/bin/feature:3:in `<top (required)>'
        11: from /usr/local/share/gems/gems/git-scripts-0.8.1/bin/feature:3:in `require_relative'
        10: from /usr/local/share/gems/gems/git-scripts-0.8.1/lib/github.rb:2:in `<top (required)>'                                                                                                  
         9: from /usr/share/rubygems/rubygems/core_ext/kernel_require.rb:59:in `require'
         8: from /usr/share/rubygems/rubygems/core_ext/kernel_require.rb:59:in `require'
         7: from /usr/local/share/gems/gems/bundler-1.17.3/lib/bundler/setup.rb:10:in `<top (required)>'                                                                                             
         6: from /usr/local/share/gems/gems/bundler-1.17.3/lib/bundler.rb:107:in `setup'
         5: from /usr/local/share/gems/gems/bundler-1.17.3/lib/bundler/runtime.rb:26:in `setup'
         4: from /usr/local/share/gems/gems/bundler-1.17.3/lib/bundler/runtime.rb:26:in `map'
         3: from /usr/share/ruby/forwardable.rb:229:in `each'
         2: from /usr/share/ruby/forwardable.rb:229:in `each'
         1: from /usr/local/share/gems/gems/bundler-1.17.3/lib/bundler/runtime.rb:31:in `block in setup'                                                                                             
/usr/local/share/gems/gems/bundler-1.17.3/lib/bundler/runtime.rb:319:in `check_for_activated_spec!': You have already activated git-scripts 0.8.1, but your Gemfile requires git-scripts 0.8.0. Prepending `bundle exec` to your command may solve this. (Gem::LoadError)
```

Not sure why that is, but I'm slightly concerned that when we try to reinstall we might run into issues like this?

**Comment by addison-grant** [comment]
CR :carryover:

**Comment by addison-grant** [comment]
CR :carryover:
QA :carryover:

**Pull request merged**